### ISSUE TITLE
[v2.5] After enabling monitoring , Pod pushprox-kube-etcd-client is in Init:CrashLoopBackOff on RancherOS

### ISSUE THREAD

**Issue by ansilh** [rancher/rancher#29815]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- Deploy a downstream cluster on RancherOS v1.5.6
- Deploy Monitoring App from Cluster Explorer 

**Result:**
pushprox-kube-etcd-client fails and goes into Init:CrashLoopBackOff

**Other details that may be helpful:**

The `metrics-cert-dir-source` volume in initContainer `copy-certs` points to  `/etc/kubernetes/ssl`
In RancherOS , the path is `/opt/rke/etc/kubernetes/ssl`.

This causes the container to fails with below

```
> kubectl get pods -A |grep -i crash
cattle-monitoring-system   pushprox-kube-etcd-client-2585j                          0/1     Init:CrashLoopBackOff   7          13m
> 
```
```
> kubectl logs pushprox-kube-etcd-client-2585j -c copy-certs -n cattle-monitoring-system 
Searching for files to copy within the source volume
cert: kube-etcd-*.pem
key: kube-etcd-*-key.pem
cacert: kube-ca.pem
Failed to find cert file
> 
```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): v2.5.1
- Installation option (single install/HA): HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported): custom
- Machine type (cloud/VM/metal) and specifications (CPU/memory): VM
- Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.13", GitCommit:"30d651da517185653e34e7ab99a792be6a3d9495", GitTreeState:"clean", BuildDate:"2020-10-15T00:59:17Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

- Docker version (use `docker version`):

```
root@ros01 ssl]# docker version
Client: Docker Engine - Community
 Version:           19.03.11
 API version:       1.40
 Go version:        go1.13.10
 Git commit:        42e35e61f3
 Built:             Mon Jun  1 09:09:53 2020
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.11
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.13.10
  Git commit:       42e35e61f3
  Built:            Mon Jun  1 09:16:24 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
[root@ros01 ssl]#
```

gzrancher/rancher#12948

**Comment by aiyengar2** [comment]
@ansilh thanks for opening this issue!

Since the certDir is an exposed field (and if all your etcd nodes are using RancherOS), the quick workaround for this should be to change the value in `rkeEtcd.clients.https.certDir` from `/etc/kubernetes/ssl` to `/opt/rke/etc/kubernetes/ssl`. You can do this on deploying / upgrading the chart by clicking on `Chart Options`, scrolling to the bottom to click on `Edit as YAML`, and modifying the YAML shown accordingly before deploying / upgrading the chart.

**Comment by jiaqiluo** [comment]
The bug is reproduced in the new Monitoring chart `9.4.201` in rancher:`v2.5-cb8a1755ccd1ed7186c24e582e6fc91fa06f87ec-head`

All cluster's nodes use RancherOS `v1.5.6`

the error shows on the `pushprox-kube-etcd-client` 
<img width="1415" alt="Screen Shot 2020-10-28 at 6 49 59 PM" src="https://user-images.githubusercontent.com/6218999/97516084-96ea9980-194f-11eb-91d9-042773515cc1.png">

@aiyengar2  the workaround seems like not working.
I upgrade the monitoring app following your suggestion 
<img width="1578" alt="Screen Shot 2020-10-28 at 6 51 32 PM" src="https://user-images.githubusercontent.com/6218999/97516191-d31dfa00-194f-11eb-9a9b-e211ea28248c.png">


Now the `pushprox-kube-etcd-client` shows the following error 
```
MountVolume.SetUp failed for volume "metrics-cert-dir-source" : hostPath type check failed: /opt/rke/etc/kubernetes/ssl is not a directory
```
<img width="1402" alt="Screen Shot 2020-10-28 at 6 55 41 PM" src="https://user-images.githubusercontent.com/6218999/97516207-dca76200-194f-11eb-8ed0-d32ed0122c6c.png">

**Comment by aiyengar2** [comment]
@jiaqiluo as indicated in https://github.com/rancher/rancher/issues/29447#issuecomment-712792964, the issue here is because of a known issue with k8s documented in https://github.com/kubernetes/kubernetes/issues/83125.

The fix is being tracked in https://github.com/rancher/charts/pull/814, so feel free to reopen the issue.

**Comment by aiyengar2** [comment]
w.r.t. the actual fix from Rancher's side to natively support monitoring RKE clusters with (some or all) nodes that use RancherOS, I'm wondering what the best way is to resolve this since PushProx does not support having >1 DaemonSets deployed for monitoring nodes with different HTTPs configurations (see below). 

TLDR: To support this, we'd either need to introduce a new exporter (`rkeRancherOSEtcd`) or modify PushProx itself to support defining multiple client configs.

#### Option 1: Add a new `rkeRancherOSEtcd` PushProx exporter

The backend changes will involve introducing a new PushProx exporter that is identical to `rkeEtcd` but sets `rkeRancherOSEtcd.clients.https.certDir=/opt/rke/etc/kubernetes/ssl`. It will also have `rkeRancherOSEtcd.enabled=false` by default.

We'll also need to make UI changes to support this for RKE clusters only, e.g.:
Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Using RancherOS for Etcd Nodes? | `N/A` | If you are using RancherOS for etcd nodes in RKE, you will need to deploy a special PushProx exporters onto your cluster | `All` or `Mixed` or `None` | `None`

If All was selected, set `rkeEtcd.enabled=false` and `rkeRancherOSEtcd.enabled=true`.

If None was selected, no changes are required for the current behavior.

If Mixed was selected, we need to expose the following options with Help Text:

`Help Text: Before deploying this chart, you will need to add labels to all of your etcd nodes that allow the PushProx clients to identify which nodes are RancherOS etcd nodes and which are non-RancherOS etcd nodes. We recommend using the labels monitoring.cattle.io/etcd-os=rancheros and monitoring.cattle.io/etcd-os=default. You will also need to ensure that any new nodes added to the cluster continue to use these labels to ensure that they are being monitored as well.`

Header / Label Name | k8s YAML | Description | Input | Default
---|---|---|---|---|
Node Selector for RancherOS Etcd Nodes | `rkeEtcd.clients.nodeSelector` | `N/A` | Ability to supply nodeSelector | `{"monitoring.cattle.io/etcd-os": "rancher-os"}`
Node Selector for Other Etcd Nodes | `rkeEtcd.clients.nodeSelector` | `N/A` | Ability to supply nodeSelector | `{"monitoring.cattle.io/etcd-os": "default"}`

#### Option 2: Modify the PushProx chart itself to support deploying multiple client DaemonSets

We could make a rather significant modification to the [`rancher-pushprox` chart](https://github.com/rancher/charts/tree/main/charts/rancher-pushprox) itself to change the `clients` field itself to support multiple different client configs (e.g. each config would correspond to a different DaemonSet deployed by the chart but all the Pods launched by the DaemonSet would still go through the same proxy).

This would still require all of the UI changes described above (with different k8s YAML fields) as well as a decently large refactor of the PushProx chart, but would avoid adding an additional RancherOS specific PushProx exporter (or any other such exporters that come in the future) and make it easier to resolve these kinds of issues in the future (e.g. supporting using a PushProx exporter for windows vs. linux nodes, which might have different scraping requirements).

It also would be better in the long-term for PushProx since it would add support for using one PushProx deployment to monitor a single component across nodes that have different scrape configuration requirements.

#### Notes

1. Regardless of which option we take, users will need to manually handle labeling nodes for setups where they are using both RancherOS and other OSs for the etcd nodes (see the Help Text for the Mixed option above); alternatively, we could choose to just disable monitoring etcd for mixed setups to avoid supporting it.
2. The only exporter this should affect are RKE etcd since no other exporter uses the `.clients.https.certDir` field at the moment.

**Comment by ansilh** [comment]
Thanks, @aiyengar2 & @jiaqiluo 
I'm able to deploy the monitoring app on RancherOS with the mentioned workaround.

- Changed `rkeEtcd.clients.https.certDir` to `/opt/rke/etc/kubernetes/ssl` and then deployed the chart
- Chart deployed successfully, but `pushprox-kube-etcd-client ` was throwing error 
- Went to the `cattle-monitoring-system` namespace and edited daemonset `pushprox-kube-etcd-client`
   Removed `type: Directory` from `hostPath` and applied the YAML (yes, this is )

I've also noticed that doing `extra_binds` of `/opt/rke/etc/kubernetes/` in kubelet fixes the problem explained in https://github.com/kubernetes/kubernetes/issues/83125

**Comment by aiyengar2** [comment]
> w.r.t. the actual fix from Rancher's side to natively support monitoring RKE clusters with (some or all) nodes that use RancherOS, I'm wondering what the best way is to resolve this since PushProx does not support having >1 DaemonSets deployed for monitoring nodes with different HTTPs configurations (see below).

The actual fix has been moved to https://github.com/rancher/rancher/issues/29840 to be tracked / prioritized separately. Users are advised to use the workaround described in https://github.com/rancher/rancher/issues/29815#issuecomment-718890687 instead.

@jiaqiluo once https://github.com/rancher/charts/pull/814 is merged in, we only need to test the workaround described in https://github.com/rancher/rancher/issues/29815#issuecomment-718283775 works.

**Comment by jiaqiluo** [comment]
The fix is validated on rancher:`v2.5-0465f049780def811e00ab32067f95734f6a88f5-head` with the lastet Monitoring chart `9.4.201`. 
All cluster's nodes use RancherOS `v1.5.6`.

The following 2 scenarios are validated:
- install monitoring v2 `9.4.201` with all default values -> wait for it fails on the `pushprox-kube-etcd-client` -> upgrade the monitoring app with the workaround -> see the problem is resolved
- install monitoring v2 `9.4.201` with the workaround -> see the app is installed and all components works
